stream/replicator.h

```diff
@@ -403,15 +464,332 @@ struct StreamReplicatorImpl {
   TerminationResponse Terminate() const { return TerminationResponse::Terminate; }
 
  private:
-  Borrowed<stream_t> stream_;
-
   // `Optional` to destroy the publisher before requesting the stream to reacquire data authority.
   Optional<Borrowed<publisher_t>> publisher_;
 };
 
 template <typename STREAM>
 using StreamReplicator = current::ss::StreamSubscriber<StreamReplicatorImpl<STREAM>, typename STREAM::entry_t>;
 
+struct MasterFlipRestrictions {
+  uint64_t max_index_diff_ = 0;
+  uint64_t max_diff_size_ = 0;
+  std::chrono::microseconds max_head_diff_ = std::chrono::microseconds(0);
+  std::chrono::microseconds max_clock_diff_ = std::chrono::microseconds(0);
+
+  MasterFlipRestrictions& SetMaxIndexDifference(uint64_t value) {
+    max_index_diff_ = value;
+    return *this;
+  }
+  MasterFlipRestrictions& SetMaxHeadDifference(std::chrono::microseconds value) {
+    max_head_diff_ = value;
+    return *this;
+  }
+  MasterFlipRestrictions& SetMaxClockDifference(std::chrono::microseconds value) {
+    max_clock_diff_ = value;
+    return *this;
+  }
+  MasterFlipRestrictions& SetMaxDiffSize(uint64_t bytes) {
+    max_diff_size_ = bytes;
+    return *this;
+  }
+};
+
+template <typename STREAM,
+          typename REPLICATOR = StreamReplicator<STREAM>,
+          ReplicationMode RM = ReplicationMode::Checked>
+class MasterFlipController final {
+ public:
+  using stream_t = STREAM;
+  using replicator_t = REPLICATOR;
+  using entry_t = typename stream_t::entry_t;
+  using publisher_t = typename stream_t::publisher_t;
+
+  MasterFlipController(Owned<STREAM>&& stream) : stream_(std::move(stream)) {}
+
+  uint64_t ExposeViaHTTP(uint16_t port,
+                         const std::string& route,
+                         MasterFlipRestrictions restrictions = MasterFlipRestrictions(),
+                         std::function<void()> flip_started = nullptr,
+                         std::function<void()> flip_finished = nullptr,
+                         std::function<void()> flip_canceled = nullptr) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (exposed_via_http_) {
+      CURRENT_THROW(StreamIsAlreadyExposedException());
+    }
+
+    exposed_via_http_ = std::make_unique<ExposedStreamState>(
+        port, route, std::move(restrictions), flip_started, flip_finished, flip_canceled);
+    exposed_via_http_->routes_scope_ +=
+        HTTP(port).Register(route, URLPathArgs::CountMask::None | URLPathArgs::CountMask::One, *Value(stream_)) +
+        HTTP(port).Register(
+            route + "/become/master", URLPathArgs::CountMask::None, [this](Request r) { MasterFlipRequest(r); });
```

Comment:
Is the construction of the URL of two pieces simply due to the fact that the `URLPathArgs::CountMask::None | URLPathArgs::CountMask::One` mask for the original subscription URL would "intercept" any call with one URL parameter?

I can see the logic, but `"/become/master"` sort of invites other `"/become/*"`, while `"/become/follower"` doesn't make immediate sense to me (and neither do any other `"/become/..."`-s). Unless there's deeper logic here, perhaps the URL would be cleaner if it's some `"/control/flip_to_master"`?

Reply:
Yes, it was the only reason of this two-pieces URL construction. `"/control/flip_to_master"` sounds less confusing for me, too.